### PR TITLE
Introduce ShardState Enum + Slight Cleanup SnapshotsInProgress

### PR DESCRIPTION
**Pull request by mkleen**
## Summary of the changes / Why this improves CrateDB

Backport of https://github.com/elastic/elasticsearch/pull/41940
Precursor for later backports.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java [hunk 1]
```diff
@@ -109,7 +106,6 @@ public Entry(Snapshot snapshot, boolean includeGlobalState, boolean partial, Sta
             } else {
                 this.shards = shards;
                 this.waitingIndices = findWaitingIndices(shards);
-                assert assertShardsConsistent(state, indices, shards);
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:112]: This method is still present in latest ES master, why did you remove it?

Comment (mkleen) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:112]: Oh, my bad. It was not present at the time of original pr https://github.com/original-brownbear/elasticsearch/blob/808b8133cf9a7baf58615f0f00ddab402e448169/server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java It must have been added later. I didn't compare to master. I will add it back.

**Comment by mkleen** [comment]
retest this please

**Comment by mkleen** [comment]
retest this please
